File: Tests/FireplaceTests.swift
import Testing

@testable import FiretermLib

@Suite("Fireplace Tests")
struct FireplaceTests {
    @Test("Initialization creates correct dimensions")
    func testInit() {
        let fp = Fireplace(width: 40, height: 15)
        #expect(fp.width == 40)
        #expect(fp.height == 15)
    }

    @Test("Update does not crash and keeps heat in bounds")
    func testUpdate() {
        var fp = Fireplace(width: 20, height: 10)
        for _ in 0..<100 {
            fp.update()
        }
        // If we got here without crashing, the simulation is stable
    }

    @Test("Render produces non-empty output")
    func testRender() {
        var fp = Fireplace(width: 20, height: 10)
        fp.update()
        var buffer = [UInt8]()
        fp.render(into: &buffer)
        #expect(buffer.count > 0)
    }

    @Test("Render contains ANSI escape codes")
    func testRenderContainsEscapes() {
        var fp = Fireplace(width: 20, height: 10)
        fp.update()
        fp.update()
        var buffer = [UInt8]()
        fp.render(into: &buffer)
        let output = String(decoding: buffer, as: UTF8.self)
        #expect(output.contains("\u{1B}["))
    }

    @Test("Multiple updates produce different frames")
    func testFramesVary() {
        var fp = Fireplace(width: 30, height: 12)

        fp.update()
        var buf1 = [UInt8]()
        fp.render(into: &buf1)

        // Run several updates to ensure state changes
        for _ in 0..<5 { fp.update() }
        var buf2 = [UInt8]()
        fp.render(into: &buf2)

        // Frames should differ due to randomness
        #expect(buf1 != buf2)
    }

    @Test("Zero-width fireplace doesn't crash")
    func testZeroWidth() {
        var fp = Fireplace(width: 0, height: 10)
        fp.update()
        var buffer = [UInt8]()
        fp.render(into: &buffer)
        #expect(buffer.count > 0)
    }

    @Test("Very small fireplace works")
    func testSmallFireplace() {
        var fp = Fireplace(width: 1, height: 1)
        for _ in 0..<10 { fp.update() }
        var buffer = [UInt8]()
        fp.render(into: &buffer)
        #expect(buffer.count > 0)
    }

    @Test("Large fireplace works")
    func testLargeFireplace() {
        var fp = Fireplace(width: 200, height: 50)
        fp.update()
        var buffer = [UInt8]()
        fp.render(into: &buffer)
        #expect(buffer.count > 0)
    }

    @Test("Render buffer can be reused")
    func testBufferReuse() {
        var fp = Fireplace(width: 20, height: 10)
        var buffer = [UInt8]()
        buffer.reserveCapacity(4096)

        fp.update()
        fp.render(into: &buffer)
        let firstLen = buffer.count

        buffer.removeAll(keepingCapacity: true)
        fp.update()
        fp.render(into: &buffer)

        // Both renders should produce meaningful output
        #expect(firstLen > 100)
        #expect(buffer.count > 100)
    }

    @Test("Heat converges towards zero at top rows after many updates")
    func testHeatDissipation() {
        var fp = Fireplace(width: 30, height: 20)
        for _ in 0..<50 { fp.update() }
        // Render and check the top portion is mostly dark (spaces)
        var buffer = [UInt8]()
        fp.render(into: &buffer)
        let output = String(decoding: buffer, as: UTF8.self)
        // The output should contain spaces (cool areas at top)
        #expect(output.contains(" "))
    }
}

File: tests/fireplace_tests.rs
use fireterm::fireplace::Fireplace;

const TEST_SIZE: (usize, usize) = (40, 80);

#[test]
fn init_creates_correct_dimensions() {
    let fp = Fireplace::new(40, 15);
    assert_eq!(fp.width, 40);
    assert_eq!(fp.height, 15);
}

#[test]
fn update_does_not_crash_and_keeps_heat_in_bounds() {
    let mut fp = Fireplace::new(20, 10);
    for _ in 0..100 {
        fp.update();
    }
}

#[test]
fn render_produces_non_empty_output() {
    let mut fp = Fireplace::new(20, 10);
    fp.update();
    let mut buffer = Vec::new();
    fp.render(&mut buffer, Some(TEST_SIZE));
    assert!(!buffer.is_empty());
}

#[test]
fn render_contains_ansi_escape_codes() {
    let mut fp = Fireplace::new(20, 10);
    fp.update();
    fp.update();
    let mut buffer = Vec::new();
    fp.render(&mut buffer, Some(TEST_SIZE));
    let output = String::from_utf8_lossy(&buffer);
    assert!(output.contains("\x1B["));
}

#[test]
fn multiple_updates_produce_different_frames() {
    let mut fp = Fireplace::new(30, 12);

    fp.update();
    let mut buf1 = Vec::new();
    fp.render(&mut buf1, Some(TEST_SIZE));

    for _ in 0..5 {
        fp.update();
    }
    let mut buf2 = Vec::new();
    fp.render(&mut buf2, Some(TEST_SIZE));

    assert_ne!(buf1, buf2);
}

#[test]
fn zero_width_fireplace_does_not_crash() {
    let mut fp = Fireplace::new(0, 10);
    fp.update();
    let mut buffer = Vec::new();
    fp.render(&mut buffer, Some(TEST_SIZE));
}

#[test]
fn very_small_fireplace_works() {
    let mut fp = Fireplace::new(1, 1);
    for _ in 0..10 {
        fp.update();
    }
    let mut buffer = Vec::new();
    fp.render(&mut buffer, Some(TEST_SIZE));
    assert!(!buffer.is_empty());
}

#[test]
fn large_fireplace_works() {
    let mut fp = Fireplace::new(200, 50);
    fp.update();
    let mut buffer = Vec::new();
    fp.render(&mut buffer, Some((60, 220)));
    assert!(!buffer.is_empty());
}

#[test]
fn render_buffer_can_be_reused() {
    let mut fp = Fireplace::new(20, 10);
    let mut buffer = Vec::with_capacity(4096);

    fp.update();
    fp.render(&mut buffer, Some(TEST_SIZE));
    let first_len = buffer.len();

    buffer.clear();
    fp.update();
    fp.render(&mut buffer, Some(TEST_SIZE));

    assert!(first_len > 100);
    assert!(buffer.len() > 100);
}

#[test]
fn heat_converges_towards_zero_at_top_rows_after_many_updates() {
    let mut fp = Fireplace::new(30, 20);
    for _ in 0..50 {
        fp.update();
    }
    let mut buffer = Vec::new();
    fp.render(&mut buffer, Some(TEST_SIZE));
    let output = String::from_utf8_lossy(&buffer);
    assert!(output.contains(' '));
}

#[test]
fn render_with_zero_terminal_size_produces_no_output() {
    let mut fp = Fireplace::new(20, 10);
    fp.update();
    let mut buffer = Vec::new();
    fp.render(&mut buffer, Some((0, 0)));
    assert!(buffer.is_empty());
}

#[test]
fn render_clamps_fire_width_to_narrow_terminal() {
    let mut fp = Fireplace::new(60, 10);
    fp.update();
    let mut buffer = Vec::new();
    fp.render(&mut buffer, Some((30, 30)));
    assert!(!buffer.is_empty());
}
